runhouse/resources/hardware/cluster.py

```diff
@@ -1957,3 +1964,195 @@ def _folder_mv(
+                den_cluster.get("status") if den_cluster.get("status") else "unknown"
+            )
+
+            # currently relying on status pings to den as a sign of cluster activity.
+            # The split is required to remove milliseconds and the offset (according to UTC) from the timestamp.
+            # (status_last_checked is in the following format: YYYY-MM-DD HH:MM:SS.ssssss±HH:MM)
+
+            last_active_at = den_cluster.get("status_last_checked")
+            last_active_at = (
+                datetime.datetime.fromisoformat(last_active_at.split(".")[0])
+                if isinstance(last_active_at, str)
+                else datetime.datetime(1970, 1, 1)
+            )  # Convert to datetime
+            last_active_at = last_active_at.replace(tzinfo=datetime.timezone.utc)
+
+            if cluster_status == "running" and not last_active_at:
+                # For BC, in case there are clusters that were saved and created before we introduced sending cluster status to den.
+                cluster_status = "unknown"
+
+            cluster_info = {
+                "Name": cluster_name,
+                "Cluster Type": cluster_type,
+                "Status": cluster_status,
+                "Last Active (UTC)": last_active_at,
+            }
+            running_clusters.append(
+                cluster_info
+            ) if cluster_status == "running" else not_running_clusters.append(
+                cluster_info
+            )
+
+        # TODO: will be used if we'll need to print not-running clusters
+
+        # Sort clusters by the 'Last Active (UTC)' and 'Status' column
+        not_running_clusters = sorted(
+            not_running_clusters,
+            key=lambda x: (x["Last Active (UTC)"], x["Status"]),
+            reverse=True,
+        )
+        running_clusters = sorted(
+            running_clusters, key=lambda x: x["Last Active (UTC)"], reverse=True
+        )
+
+        return running_clusters, not_running_clusters
+
+    ###############################
+    # Cluster list
+    ###############################
+    @staticmethod
+    def list(
+        get_all_clusters: Optional[bool] = False,
+        since: Optional[bool] = None,
+        status: Optional[Union[str, ResourceServerStatus]] = None,
+    ):
+
+        import sky
+
+        cluster_filters = (
+            Cluster._parse_filters(since=since, cluster_status=status)
+            if not get_all_clusters
+            else {"all": "all"}
+        )
+
+        # get clusters from den
+        den_clusters_resp = Cluster.get_clusters_from_den(
+            cluster_filters=cluster_filters
+        )
+        if den_clusters_resp.status_code != 200:
+            logger.error(f"Failed to load {rns_client.username}'s clusters from Den")
+            den_clusters = []
+        else:
+            den_clusters = den_clusters_resp.json().get("data")
+
+        # get sky live clusters
+        sky_live_clusters = sky.status(refresh=True)
```

Comment:
just confirming - does this by default only show live ones? (what about ones in "INIT" state for example?)

Reply:
According to skypilot [documentation](https://skypilot.readthedocs.io/en/latest/reference/api.html#sky-status), `sky.status()` will return clusters with the following statuses: `UP`, `INIT` or `STOPPED`. 

If the cluster is saved in den, we won't treat it as sky cluster, even if its status is `INIT` or `STOPPED`. 

It’s more of a product question, but we’ll probably need to decide how to handle `INIT` clusters that are saved to the den. I think it might be worth consulting with Donny about this because I remember him mentioning that we want to minimize the use of sky commands and create a wrapper around them, so users will interact with the Runhouse API instead of the Skypilot API.